example_client/multi_inputs.py

```diff
@@ -0,0 +1,280 @@
+      if not ret:
+        cam.set(cv2.CAP_PROP_POS_FRAMES, 0)
+        continue
+      # normalize the video frame dimension to that of the camera
+      else:
+        # to maintain the frame inferencing parity with the cameras, lets sleep
+        # here to maintain cam_fps speed
+        sleep((1000 / CAM_FPS) / 1000)
+        frame = cv2.resize(frame, (CAM_WIDTH, CAM_HEIGHT))
+
+    fps = fps + 1
+    if (time() - cam_start_time) * 1000 >= 1000:
+      log.warning('{}{} fps: {}, Inf fps: {}, dropped fps: {}'
+                  .format(src_type, thr_id, fps, inf_fps, dropped_fps))
+      fps = 0
+      inf_fps = 0
+      dropped_fps = 0
+      cam_start_time = time()
+
+    # resize the frame to what network input layer expects it to be
+    image = cv2.resize(frame, (input_dimension, input_dimension))
+    image = image.transpose(2, 0, 1).reshape(1, 3, input_dimension, input_dimension)
+
+    inf_time = time()
+    # protobuf the input
+    request.inputs[input_layer].CopyFrom(
+        tf_contrib_util.make_tensor_proto(image, shape=None))
+
+    thr_lock.acquire()
+    try:
+      result = stub.Predict(request, 10.0)
+    except Exception as e:
+      log.error('Caught exception {}'.format(e))
+      cam.release()
+      return
+    thr_lock.release()
+    duration = time() - inf_time
+
+    # we have to decode the received protobuf format
+    res = tf_contrib_util.make_ndarray(result.outputs[output_layer])
+
+    if not res.any():
+      log.error('Thr{}: Predictions came back with wrong output layer name'.format(thr_id))
+      dropped_fps = dropped_fps + 1
+      disp_buf[thr_id] = frame
+    else:
+      log.debug('Predictions came back fine')
+      inf_fps = inf_fps + 1
+      disp_buf[thr_id] = parse_output(thr_id, res, frame)
+
+  # while exit_ok == False
+
+  cam.release()
+  log.warning('Exiting thread {}'.format(thr_id))
+
+#####################################################################################
+
+def main():
+  log.basicConfig(format="[$(levelname)s ] %(message)s", level=log.INFO, stream=sys.stdout)
+  args = build_argparser().parse_args()
+  num_cam = args.num_cameras if (args.num_cameras) else 0
+  vid_src = args.file
+  network_name = args.network_name
+  input_layer = args.input_layer
+  output_layer = args.output_layer
+  input_dimension = args.frame_size
+
+  if not args.file and not args.num_cameras:
+    log.error('Pls supply either the camera or the video file. Try -f for options')
+    return
+
+  video_files = []
+  if vid_src:
+    if os.path.isdir(vid_src):
+      for r, d, f in os.walk(vid_src):
+        for f_ in f:
+          # only mp4 files supported as of now
+          if '.mp4' in f_:
+            video_files.append(r + f_)
+    elif os.path.isfile(vid_src):
+      if '.mp4' in vid_src:
+        video_files.append(vid_src)
+
+  # ovms connection
+  channel = grpc.insecure_channel("{}:{}".format(args.ip, args.port))
+  stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)
```

Comment:
@mareklevv correct me if I'm wrong, but wouldn't it be better to have `stub` (or event `channel` as well) created per thread?
From: https://www.grpc.io/docs/guides/concepts/
"On the client side, the client has a local object known as stub (for some languages, the preferred term is client)"

I'm thinking of taking the most out of upcoming multiple worker release.

Reply:
makes sense. modified code to create local stub per thread